AlphaWallet/Core/Enjin/EnjinProvider.swift

```diff
@@ -0,0 +1,448 @@
+
+        if fetch.isResolved {
+            fetch = Promise<[AlphaWallet.Address: [GetEnjinBalancesQuery.Data.EnjinBalance]]> { seal in
+                let offset = 1
+                fetchPage(forOwner: owner, offset: offset) { result in
+                    switch result {
+                    case .success(let result):
+                        seal.fulfill(result)
+                    case .failure(let error):
+                        seal.reject(error)
+                    }
+                }
+            }.then({ balances -> Promise<[AlphaWallet.Address: [GetEnjinTokenQuery.Data.EnjinToken]]> in
+                let ids = (balances[owner] ?? []).compactMap { $0.token?.id }
+                return EnjinProvider.functional.getTokens(graphqlClient: self.graphqlClient, ids: ids, owner: owner)
+            })
+        }
+        return fetch
+    }
+
+    private func fetchPage(forOwner owner: AlphaWallet.Address, offset: Int, sum: [AlphaWallet.Address: [GetEnjinBalancesQuery.Data.EnjinBalance]] = [:], completion: @escaping (Swift.Result<[AlphaWallet.Address: [GetEnjinBalancesQuery.Data.EnjinBalance]], OpenSeaError>) -> Void) {
+
+        graphqlClient.fetch(query: GetEnjinBalancesQuery(ethAddress: owner.eip55String, page: offset, limit: 50)) { [weak self] response in
+            guard let strongSelf = self else { return }
+
+            switch response {
+            case .failure(let error):
+                completion(.failure(OpenSeaError(localizedDescription: "Error calling Engin API: \(String(describing: error))")))
+            case .success(let graphQLResult):
+                var results = sum
+                let balances: [GetEnjinBalancesQuery.Data.EnjinBalance]
+                if let data = graphQLResult.data {
+                    balances = (data.enjinBalances ?? []).compactMap { $0 }
+                } else {
+                    balances = []
+                }
+
+                if let values = results[owner] {
+                    results[owner] = values + balances
+                } else {
+                    results[owner] = balances
+                }
+
+                if !balances.isEmpty {
+                    strongSelf.fetchPage(forOwner: owner, offset: offset + 1, sum: results) { results in
+                        completion(results)
+                    }
+                } else {
+                    let excludingUefa = sum
+                    var tokenIdCount = 0
+                    for (_, tokenIds) in excludingUefa {
+                        tokenIdCount += tokenIds.count
+                    }
+                    strongSelf.cachePromise(withTokenIdCount: tokenIdCount, forOwner: owner)
+
+                    completion(.success(excludingUefa))
+                }
+            }
+        }
+    }
+
+    private func cachePromise(withTokenIdCount tokenIdCount: Int, forOwner wallet: AlphaWallet.Address) {
+        guard tokenIdCount >= EnjinProvider.numberOfTokenIdsBeforeRateLimitingRequests else { return }
```

Comment:
This just blocks subsequent fetches, but doesn't rate limit, right?

Reply:
not quite sure, that was already presented in OpenSea, haven't check what is really does